af/analysis/ooni_db_backup.py

```diff
@@ -0,0 +1,287 @@
+    colnames = []
+    blob = []
+    for row in click.execute(f"DESCRIBE TABLE {tblname}"):
+        colname, coltype = row[:2]
+        colnames.append(colname)
+        blob.append(f"{colname} {coltype}")
+
+    assert colnames
+    s_blob = ", ".join(blob)  # "measurement_uid String report_id String, ... "
+    s_colnames = ", ".join(colnames)  # "measurement_uid, report_id, ... "
+    return s_colnames, s_blob
+
+
+def export_fastpath(click, aws_id: str, aws_secret: str, baseurl: str) -> None:
+    tblname = "fastpath"
+    s_colnames, s_blob = describe_table(click, tblname)
+    assert s_blob.startswith("measurement_uid String, ")
+    # s_colnames is a string to be placed into ##colnames## without quotes
+    basesql = """
+    INSERT INTO FUNCTION s3(%(s3path)s, %(aws_id)s, %(aws_secret)s,
+        'CSVWithNames',
+        %(s_blob)s
+    )
+    SELECT ##colnames##
+    FROM fastpath
+    WHERE toYYYYMM(measurement_start_time) = %(yyyymm)s
+    SETTINGS s3_max_connections=4, max_insert_threads=4,
+    s3_min_upload_part_size=50100100;
+    """
+    delta_ms = 0
+    for year in range(2012, 2024):
+        for month in range(1, 13):
+            yyyymm = f"{year}{month:02}"
+            s3path = f"{baseurl}/csv/{tblname}_{yyyymm}.csv.zstd"
+            kw = dict(
+                aws_id=aws_id,
+                aws_secret=aws_secret,
+                s3path=s3path,
+                yyyymm=yyyymm,
+                s_blob=s_blob,
+            )
+            sql = basesql.replace("##colnames##", s_colnames)
+            log.info(f"exporting {tblname} {yyyymm}")
+            t0 = perf_counter()
+            log.info(sql)
+            click.execute(sql, kw)
+            delta_ms += int((perf_counter() - t0) * 1000)
+            metrics.gauge(f"table_{tblname}_backup_time_ms", delta_ms)
+            sleep(10)
+
+    log.info("table backup completed")
+
+
+def export_jsonl(click, aws_id: str, aws_secret: str, baseurl: str) -> None:
+    tblname = "jsonl"
+    s_colnames, s_blob = describe_table(click, tblname)
+    assert s_blob.startswith("report_id String, ")
+    # s_colnames is a string to be placed into ##colnames## without quotes.
+    # jsonl is indexed by report_id: chuck using LIMIT / OFFSET instead
+    basesql = """
+    INSERT INTO FUNCTION s3(%(s3path)s, %(aws_id)s, %(aws_secret)s,
+        'CSVWithNames',
+        %(s_blob)s
+    )
+    SELECT ##colnames##
+    FROM jsonl
+    SETTINGS s3_max_connections=4, max_insert_threads=4,
+    s3_min_upload_part_size=50100100;
+    """
+    delta_ms = 0
+    chunk_size = 100_000_000
+    chunk_num = 0
+    while True:
+        s3path = f"{baseurl}/csv/{tblname}_{chunk_num:05}.csv.zstd"
+        kw = dict(
+            aws_id=aws_id,
+            aws_secret=aws_secret,
+            s3path=s3path,
+            limit=chunk_size,
+            offset=chunk_num * chunk_size,
+            s_blob=s_blob,
+        )
+        sql = "SELECT count() FROM jsonl LIMIT %(limit)s OFFSET %(offset)s"
```

Comment:
If I understand the intent of this function correctly, you are trying to export the jsonl table in several chunks of 100M rows and use this `count()` query to understand if you are at the last chunk.

I believe there are however two bugs atm:

1. the `basesql` query does not make use of `LIMIT` and `OFFSET` so it's effectively just selecting and copying the whole table on the first iteration of the `while True` loop
2. the above `SELECT count()` query will always return empty rows on the second interation, because `LIMIT` and `OFFSET` constraints are being applied on the result of the `count()` query.

If you want to know the count of the current chunk, you need an extra level of nesting in the SELECT query:
```
SELECT COUNT() FROM (
    SELECT measurement_uid FROM jsonl LIMIT %(limit)s OFFSET %(offset)s
)
```

That said, I'm not sure this is the best approach, as on every iteration you are basically scanning all the rows until the offset at every iteration of the loop.

It's probably better to make use of the `PARTITION BY` clause, which I would suggest doing on the `date` (AFAIK `date` is what we used to call `bucket_date`) column which we know is going to always increase and will not contain new data going forward.

Reply:
That's correct. The function was still being finished and tested. Now it's rewritten to use a temporary file due to a bug in Clickhouse in the uploading to S3.